src/Evolution/DiscontinuousGalerkin/SlopeLimiters/HwenoModifiedSolution.hpp

```diff
@@ -0,0 +1,494 @@
+        }
+        return result;
+      }
+      ();
+      b += quadrature_weights_dot_u *
+           quadrature_weights_dot_interpolation_matrix;
+    }
+  }
+  return b;
+}
+
+// Solve the constrained fit problem that gives the HWENO modified solution,
+// for one particular tensor component. For details, see documentation of
+// `compute_hweno_modified_neighbor_solution` below.
+template <typename Tag, size_t VolumeDim, typename Package>
+void solve_constrained_fit(
+    const gsl::not_null<DataVector*> constrained_fit_result,
+    const DataVector& u, const size_t tensor_index,
+    const Element<VolumeDim>& element, const Mesh<VolumeDim>& mesh,
+    const std::unordered_map<
+        std::pair<Direction<VolumeDim>, ElementId<VolumeDim>>, Package,
+        boost::hash<std::pair<Direction<VolumeDim>, ElementId<VolumeDim>>>>&
+        neighbor_data,
+    const std::pair<Direction<VolumeDim>, ElementId<VolumeDim>>&
+        primary_neighbor,
+    const std::vector<std::pair<Direction<VolumeDim>, ElementId<VolumeDim>>>&
+        neighbors_to_exclude) noexcept {
+  // Get the cache of linear algebra quantities for this element
+  const ConstrainedFitCache<VolumeDim>& cache =
+      constrained_fit_cache(element, mesh);
+
+  // Because we don't support h-refinement, the direction is the only piece
+  // of the neighbor information that we actually need.
+  const Direction<VolumeDim> primary_direction = primary_neighbor.first;
+  const std::vector<Direction<VolumeDim>>
+      directions_to_exclude = [&neighbors_to_exclude]() noexcept {
+    std::vector<Direction<VolumeDim>> result(neighbors_to_exclude.size());
+    for (size_t i = 0; i < result.size(); ++i) {
+      result[i] = neighbors_to_exclude[i].first;
+    }
+    return result;
+  }
+  ();
+
+  const DataVector& w = cache.quadrature_weights;
+  const DirectionMap<VolumeDim, Matrix>& m = cache.interpolation_matrices;
+  const DirectionMap<VolumeDim, DataVector>& w_dot_m =
+      cache.quadrature_weights_dot_interpolation_matrices;
+
+  // Use cache if possible, or compute matrix if we are in edge case
+  const Matrix& a_inverse = LIKELY(directions_to_exclude.size() < 2)
+                                ? cache.retrieve_a_inverse_matrix(
+                                      primary_direction, directions_to_exclude)
+                                : Hweno_detail::a_inverse_matrix(
+                                      element, mesh, w, m, w_dot_m,
+                                      primary_direction, directions_to_exclude);
+
+  const DataVector b =
+      b_vector<Tag>(mesh, tensor_index, w, m, w_dot_m, neighbor_data,
+                    primary_neighbor, neighbors_to_exclude);
+
+  const size_t number_of_points = b.size();
+  const DataVector a_inverse_b = apply_matrices(
+      std::array<Matrix, 1>{{a_inverse}}, b, Index<1>(number_of_points));
+  const DataVector a_inverse_w = apply_matrices(
+      std::array<Matrix, 1>{{a_inverse}}, w, Index<1>(number_of_points));
```

Comment:
Use reference_wrappers to avoid copying the matrices here.

Reply:
To make sure I understand: are you suggesting I write the first argument passed into `apply_matrices` as `std::array<std::reference_wrapper<Matrix>, 1>{{a_inverse}}` ?

Reply:
Yes.  (Might need to be `std::reference_wrapper<const Matrix>`.)